File: sdk-go/images.go
package boxkite

import (
	"context"
	"fmt"
	"net/url"
)

// CreateImageRequest is the request body for CreateImage
// (POST /v1/images). All fields are optional -- a zero-value
// CreateImageRequest{} builds an unmodified "boxkite-default" image.
type CreateImageRequest struct {
	// Label is an optional human-readable label for the image.
	Label *string `json:"label,omitempty"`
	// Base is one of "boxkite-default", "boxkite-minimal", "boxkite-node",
	// "boxkite-go". Defaults to "boxkite-default" server-side when empty.
	// "boxkite-node" drops Python entirely (no PythonPackages installable,
	// only AptPackages/NpmPackages); "boxkite-go" drops both Python and
	// Node entirely (no PythonPackages or NpmPackages installable, only
	// AptPackages).
	Base string `json:"base,omitempty"`
	// PythonPackages are exact-version-pinned packages ("name==version",
	// no ranges) to install into the image.
	PythonPackages []string `json:"python_packages,omitempty"`
	// AptPackages are exact-version-pinned apt packages ("name==version",
	// no ranges) to install into the image.
	AptPackages []string `json:"apt_packages,omitempty"`
	// NpmPackages are exact-version-pinned npm packages ("name==version"
	// or "@scope/name==version", no ranges) to install into the image. Not
	// supported on Base="boxkite-go".
	NpmPackages []string `json:"npm_packages,omitempty"`
}

// Image is a custom sandbox image, as returned by CreateImage/GetImage/
// ListImages.
type Image struct {
	ID             string         `json:"id"`
	Label          *string        `json:"label"`
	Base           string         `json:"base"`
	PythonPackages []string       `json:"python_packages"`
	AptPackages    []string       `json:"apt_packages"`
	NpmPackages    []string       `json:"npm_packages"`
	Status         string         `json:"status"`
	Digest         *string        `json:"digest"`
	RegistryRef    *string        `json:"registry_ref"`
	ScanResult     map[string]any `json:"scan_result"`
	FailureReason  *string        `json:"failure_reason"`
	CreatedAt      string         `json:"created_at"`
	CompletedAt    *string        `json:"completed_at"`
}

// CreateImage queues a build of a custom sandbox image
// (POST /v1/images). Always asynchronous -- returns immediately with
// Status "queued"; poll GetImage for progress. 404s ("feature_disabled")
// if the deployment hasn't enabled the declarative builder
// (BOXKITE_IMAGE_BUILDER_ENABLED).
func (c *Client) CreateImage(ctx context.Context, req CreateImageRequest) (*Image, error) {
	var out Image
	if err := c.doJSON(ctx, "POST", "/v1/images", req, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetImage fetches one custom sandbox image's build status and details
// (GET /v1/images/{id}).
func (c *Client) GetImage(ctx context.Context, imageID string) (*Image, error) {
	var out Image
	path := fmt.Sprintf("/v1/images/%s", url.PathEscape(imageID))
	if err := c.doJSON(ctx, "GET", path, nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListImages lists every custom sandbox image built for this account
// (GET /v1/images).
func (c *Client) ListImages(ctx context.Context) ([]Image, error) {
	var out []Image
	if err := c.doJSON(ctx, "GET", "/v1/images", nil, &out, nil); err != nil {
		return nil, err
	}
	if out == nil {
		out = []Image{}
	}
	return out, nil
}

// DeleteImage deletes a custom sandbox image (DELETE /v1/images/{id}).
func (c *Client) DeleteImage(ctx context.Context, imageID string) error {
	path := fmt.Sprintf("/v1/images/%s", url.PathEscape(imageID))
	return c.doJSON(ctx, "DELETE", path, nil, nil, nil)
}

File: sdk-rust/src/images.rs
//! Declarative builder (custom sandbox images): `POST/GET/DELETE
//! /v1/images*`. Mirrors `sdk-python`'s `create_image`/`get_image`/
//! `list_images`/`delete_image`.

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::client::Client;
use crate::error::BoxkiteError;

/// Pre-approved base image for [`Client::create_image`]. Each variant is
/// itself a separate, boxkite-maintained hardened image -- never a
/// caller-supplied arbitrary base OS.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum ImageBase {
    /// The full data-science/document/browser stack.
    #[serde(rename = "boxkite-default")]
    BoxkiteDefault,
    /// A lean Python+Node base with nothing preinstalled.
    #[serde(rename = "boxkite-minimal")]
    BoxkiteMinimal,
    /// Drops Python entirely -- no `python_packages` installable.
    #[serde(rename = "boxkite-node")]
    BoxkiteNode,
    /// Drops both Python and Node entirely -- no `python_packages` or
    /// `npm_packages` installable.
    #[serde(rename = "boxkite-go")]
    BoxkiteGo,
}

/// Builder for `POST /v1/images`'s optional fields.
#[derive(Debug, Clone, Default, Serialize)]
pub struct CreateImageOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    base: Option<ImageBase>,
    #[serde(skip_serializing_if = "Option::is_none")]
    python_packages: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    apt_packages: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    npm_packages: Option<Vec<String>>,
}

impl CreateImageOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn label(mut self, label: impl Into<String>) -> Self {
        self.label = Some(label.into());
        self
    }

    /// Defaults to [`ImageBase::BoxkiteDefault`] server-side when omitted.
    pub fn base(mut self, base: ImageBase) -> Self {
        self.base = Some(base);
        self
    }

    /// Exact-version-pinned packages (`"name==version"`, no ranges), e.g.
    /// `"polars==1.9.0"`. Not supported on [`ImageBase::BoxkiteNode`] or
    /// [`ImageBase::BoxkiteGo`].
    pub fn python_packages<I, S>(mut self, packages: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.python_packages = Some(packages.into_iter().map(Into::into).collect());
        self
    }

    /// Exact-version-pinned apt/apk packages (`"name==version"`, no ranges).
    pub fn apt_packages<I, S>(mut self, packages: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.apt_packages = Some(packages.into_iter().map(Into::into).collect());
        self
    }

    /// Exact-version-pinned npm packages (`"name==version"`, no ranges), e.g.
    /// `"typescript==5.6.0"`. Not supported on [`ImageBase::BoxkiteGo`].
    pub fn npm_packages<I, S>(mut self, packages: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.npm_packages = Some(packages.into_iter().map(Into::into).collect());
        self
    }
}

/// A custom sandbox image (`SandboxImageOut`/`SandboxImageBuildAccepted`).
/// [`Client::create_image`]'s response only ever populates `id`/`label`/
/// `status`/`created_at` -- the rest fill in once you poll
/// [`Client::get_image`] and the build reaches a terminal state.
#[derive(Debug, Clone, Deserialize)]
pub struct Image {
    pub id: String,
    pub label: Option<String>,
    #[serde(default)]
    pub base: String,
    #[serde(default)]
    pub python_packages: Vec<String>,
    #[serde(default)]
    pub apt_packages: Vec<String>,
    #[serde(default)]
    pub npm_packages: Vec<String>,
    /// `"queued"`, `"building"`, `"scanning"`, `"completed"`, `"failed"`, or
    /// `"rejected"`.
    pub status: String,
    pub digest: Option<String>,
    pub registry_ref: Option<String>,
    pub scan_result: Option<serde_json::Value>,
    pub failure_reason: Option<String>,
    pub created_at: String,
    pub completed_at: Option<String>,
}

impl Client {
    /// `POST /v1/images` -- build a custom sandbox image (a base image plus
    /// exact-version-pinned Python/apt/npm packages). Returns immediately
    /// with a pending record; poll [`Client::get_image`] for `status` to
    /// reach a terminal state.
    pub async fn create_image(&self, options: CreateImageOptions) -> Result<Image, BoxkiteError> {
        let builder = self.request(Method::POST, "/v1/images").json(&options);
        self.send(builder).await
    }

    /// `GET /v1/images/{id}` -- a custom sandbox image's build status and details.
    pub async fn get_image(&self, image_id: &str) -> Result<Image, BoxkiteError> {
        let builder = self.request(Method::GET, &format!("/v1/images/{image_id}"));
        self.send(builder).await
    }

    /// `GET /v1/images` -- every custom sandbox image built for this account.
    pub async fn list_images(&self) -> Result<Vec<Image>, BoxkiteError> {
        let builder = self.request(Method::GET, "/v1/images");
        self.send_or_default(builder).await
    }

    /// `DELETE /v1/images/{id}` -- delete a custom sandbox image.
    pub async fn delete_image(&self, image_id: &str) -> Result<(), BoxkiteError> {
        let builder = self.request(Method::DELETE, &format!("/v1/images/{image_id}"));
        self.send_no_content(builder).await
    }
}
